# EXPLORACION Y LIMPIEZA DE LOS DATASETS 

In [1]:
import pandas as pd
import numpy as np
pd.set_option('display.max_colwidth', None)
pd.set_option('display.max_rows',None)
pd.set_option('display.max_columns',None)

Recuperamos los datasets guardados en formato CSV :

In [2]:
df_Alquiler=pd.read_csv('df_Alquiler.csv')
df_Comprar=pd.read_csv('df_Comprar.csv')
Alquiler_Parking=pd.read_csv('Alquiler_Parking.csv')
Comprar_Parking=pd.read_csv('Comprar_Parking.csv')
Alquiler_Amueblado=pd.read_csv('Alquiler_Amueblado.csv')

## CREACION DEL DATASET DE VIVIENDAS A LA VENTA

## Creamos un 'id' único

Partimos de la base que la variable 'id' que representa cada anuncio debería ser única :

In [3]:
df_Comprar.shape

(18753, 16)

In [4]:
df_Comprar.nunique()

Unnamed: 0     18753
id             18718
url            18753
subtype           12
transaction        1
owner              2
sell-type          1
provider        1798
title          14417
location          74
features       10556
description    15114
price           4023
price2          1572
m2               490
rooms             20
dtype: int64

De un total de 18753, solo tenemos 18718 id únicos de manera que 35 son duplicados :

In [5]:
df_Comprar[df_Comprar.duplicated(subset='id',keep=False)]

Unnamed: 0                id  \
359           359   id2319004426926   
360           360   id2319004426926   
3351         3351   id4975004332920   
3352         3352   id4975004332920   
3366         3366   id4975004332050   
3367         3367   id4975004332050   
3846         3846  id28584000000672   
3847         3847  id44769000000018   
3848         3848  id14774004265791   
3850         3850  id45461000000022   
3852         3852   id4737003812364   
3855         3855  id44555000000188   
3857         3857   id1934003602726   
3858         3858  id13478004396395   
3859         3859  id28584000000672   
3861         3861   id5329004352106   
3863         3863  id44555000000188   
3864         3864  id44769000000018   
3865         3865  id45461000000022   
3867         3867   id1934003602726   
3870         3870   id4737003812364   
3871         3871   id5329004352106   
3873         3873  id14774004265791   
3874         3874  id13478004396395   
4926         4926  id26096004027783   
4927         4927  id26096004027783   
7014         7014    id885003214967   
7030         7030    id885003214298   
7083         7083    id885003214302   
7319         7319  id21820000003750   
7320         7320  id21820000003750   
9212         9212  id23591004307472   
9219         9219  id23591004307472   
9261         9261  id13478004396258   
9265         9265  id13478004396258   
12978       12978   id8166003786000   
12979       12979   id8166003786000   
13037       13037  id16510003356181   
13039       13039  id16510003356181   
13044       13044  id44250000001021   
13047       13047  id23591004307713   
13054       13054  id23591004307713   
13057       13057  id44250000001021   
13083       13083   id2861002717598   
13084       13084   id2861002717598   
15420       15420   id8478003867827   
15422       15422   id8478003867827   
16126       16126   id5329004353655   
16127       16127   id5329004353655   
17628       17628   id1507004169354   
17631       17631   id1507004169354   
17687       17687   id8166003791015   
17689       17689   id8166003788348   
17690       17690   id8166003788348   
17691       17691   id8166003791015   
18098       18098   id5612002808374   
18104       18104   id8166003788876   
18105       18105   id8166003789428   
18110       18110   id5612002808374   
18111       18111   id8166003789428   
18112       18112   id8166003788876   
18123       18123  id10944000008274   
18125       18125  id10944000008274   
18189       18189    id885003214967   
18206       18206    id885003214298   
18334       18334    id885003214302   
18335       18335    id885003214302   
18348       18348   id4017003409634   
18351       18351   id4017003409634   

                                                                                                                                                                                                                                  url  \
359                                                                                                https://www.habitaclia.com/comprar-piso-a_reformar_raval-barcelona-i2319004426926.htm?pag=23&f=&geo=d&coddists=100&from=list&lo=55   
360                                                                                                https://www.habitaclia.com/comprar-piso-a_reformar_raval-barcelona-i2319004426926.htm?pag=24&f=&geo=d&coddists=100&from=list&lo=55   
3351                                https://www.habitaclia.com/comprar-piso-en_venta_espectacular_en_una_finca_modernista_en_la_calle_a_dreta_de_l_eixample-barcelona-i4975004332920.htm?pag=53&f=&geo=d&coddists=200&from=list&lo=55   
3352                                https://www.habitaclia.com/comprar-piso-en_venta_espectacular_en_una_finca_modernista_en_la_calle_a_dreta_de_l_eixample-barcelona-i4975004332920.htm?pag=54&f=&geo=d&coddists=200&from=list&lo=55   
3366                      https://www.habitaclia.com/comprar-atico-reformado_a_estrenar_en_enric_granados_una_d

Al constatar que se tratan de anuncios repetidos, eliminamos duplicados y nos quedamos con la primera ocurrencia :

In [6]:
df_Comprar.drop_duplicates(subset=['id'], keep='first',inplace=True)

In [7]:
df_Comprar.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18718 entries, 0 to 18752
Data columns (total 16 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   18718 non-null  int64 
 1   id           18718 non-null  object
 2   url          18718 non-null  object
 3   subtype      18718 non-null  object
 4   transaction  18718 non-null  object
 5   owner        18718 non-null  object
 6   sell-type    18718 non-null  object
 7   provider     18718 non-null  object
 8   title        18718 non-null  object
 9   location     18718 non-null  object
 10  features     18717 non-null  object
 11  description  18718 non-null  object
 12  price        18688 non-null  object
 13  price2       18718 non-null  int64 
 14  m2           18718 non-null  int64 
 15  rooms        18718 non-null  int64 
dtypes: int64(4), object(12)
memory usage: 2.4+ MB


In [8]:
Comprar_Parking.shape

(2011, 15)

In [9]:
Comprar_Parking.nunique()

id             2010
url            2011
subtype          11
transaction       1
owner             2
sell-type         1
provider        546
title          1657
location         70
features       1552
description    1608
price           912
price2          559
m2              390
rooms            16
dtype: int64

In [10]:
Comprar_Parking.drop_duplicates(subset=['id'], keep='first',inplace=True)

In [11]:
Comprar_Parking.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 2010 entries, 0 to 2010
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   id           2010 non-null   object
 1   url          2010 non-null   object
 2   subtype      2010 non-null   object
 3   transaction  2010 non-null   object
 4   owner        2010 non-null   object
 5   sell-type    2010 non-null   object
 6   provider     2010 non-null   object
 7   title        2010 non-null   object
 8   location     2010 non-null   object
 9   features     2010 non-null   object
 10  description  2010 non-null   object
 11  price        2002 non-null   object
 12  price2       2010 non-null   int64 
 13  m2           2010 non-null   int64 
 14  rooms        2010 non-null   int64 
dtypes: int64(3), object(12)
memory usage: 251.2+ KB


## Incorporamos variable parking Si/No

Usamos la función assign() para incorporar la variable parking en el dataset :

In [12]:
df_Comprar=df_Comprar.assign(parking=df_Comprar.id.isin(Comprar_Parking.id).astype(int))

In [13]:
df_Comprar['parking']=np.where(df_Comprar['parking']==1,'Si','No')

In [14]:
df_Comprar['parking'].value_counts()

No    16723
Si     1995
Name: parking, dtype: int64

In [15]:
df_Comprar.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18718 entries, 0 to 18752
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   18718 non-null  int64 
 1   id           18718 non-null  object
 2   url          18718 non-null  object
 3   subtype      18718 non-null  object
 4   transaction  18718 non-null  object
 5   owner        18718 non-null  object
 6   sell-type    18718 non-null  object
 7   provider     18718 non-null  object
 8   title        18718 non-null  object
 9   location     18718 non-null  object
 10  features     18717 non-null  object
 11  description  18718 non-null  object
 12  price        18688 non-null  object
 13  price2       18718 non-null  int64 
 14  m2           18718 non-null  int64 
 15  rooms        18718 non-null  int64 
 16  parking      18718 non-null  object
dtypes: int64(4), object(13)
memory usage: 2.6+ MB


In [16]:
df_Comprar.head()

Unnamed: 0                id  \
0           0  id41022004397639   
1           1   id4190004382165   
2           2  id41022004405185   
3           3  id29169000001903   
4           4  id41022004397648   

                                                                                                                                                                    url  \
0                           https://www.habitaclia.com/comprar-piso-obra_nueva_calle_carme_23_raval-barcelona-i41022004397639.htm?f=&geo=d&coddists=100&from=list&lo=55   
1          https://www.habitaclia.com/comprar-piso-oportunidad_de_inversion_en_pleno_centro_de_raval-barcelona-i4190004382165.htm?f=&geo=d&coddists=100&from=list&lo=55   
2                https://www.habitaclia.com/comprar-piso-obra_nueva_calle_princep_de_viana_14_raval-barcelona-i41022004405185.htm?f=&geo=d&coddists=100&from=list&lo=55   
3                                                     https://www.habitaclia.com/comprar-piso-gotic-barcelona-i29169000001903.htm?f=&geo=d&coddists=100&from=list&lo=55   
4  https://www.habitaclia.com/comprar-piso-obra_nueva_calle_massanet_6_st_pere_sta_caterina_el_born-barcelona-i41022004397648.htm?f=&geo=d&coddists=100&from=list&lo=55   

  subtype transaction         owner    sell-type provider  \
0    FLAT        SALE  PROFESSIONAL  SECOND_HAND  41022-1   
1    FLAT        SALE  PROFESSIONAL  SECOND_HAND   4190-2   
2    FLAT        SALE  PROFESSIONAL  SECOND_HAND  41022-1   
3    FLAT        SALE  PROFESSIONAL  SECOND_HAND  29169-1   
4    FLAT        SALE  PROFESSIONAL  SECOND_HAND  41022-1   

                                                                      title  \
0                                      Piso en Carrer carme, 23. Obra nueva   
1  Piso Riera baixa. Oportunidad de inversión en pleno centro de barcelona.   
2                           Piso en Carrer princep de viana, 14. Obra nueva   
3                                                      Piso Carrer d'avinyó   
4                                    Piso en Carrer massanet, 6. Obra nueva   

                                         location  \
0                               Barcelona - Raval   
1                               Barcelona - Raval   
2                               Barcelona - Raval   
3                               Barcelona - Gòtic   
4  Barcelona - St. Pere - Sta. Caterina - El Born   

                                       features  \
0      61m2 - 1 habitacion - 1 baño - 4.754€/m2   
1    37m2 - 2 habitaciones - 1 baño - 2.919€/m2   
2    60m2 - 2 habitaciones - 1 baño - 6.267€/m2   
3  101m2 - 3 habitaciones - 2 baños - 7.079€/m2   
4   82m2 - 2 habitaciones - 2 baños - 4.634€/m2   

                                                                                                                   description  \
0  En el distrito de Ciutat Vella, en pleno barrio del Raval, les ofrecemos una promoción de 10 viviendas en un edificio de...   
1    Oportunidad de inversión en pleno centro de Barcelona. Vivienda de 37m² que cuenta con dos habitaciones y un baño y un...   
2  Exclusiva promoción de 11 viviendas únicas de obra nueva. Se trata de la rehabilitación integral de una finca clásica, e...   
3   En el emblemático barrio del Gotic , el barrio con más personalidad de la ciudad Condal delimitado por Las Ramblas, Vía...   
4   En el emblemático barrio del Born, junto al mercado de Santa Caterina y próximo a la Catedral de Barcelona, les ofrecem...   

                                        price  price2   m2  rooms parking  
0  Oportunidad\n290.000 €\nha bajado 10.000 €  290000   61      1      No  
1                      Oportunidad\n108.000 €  108000   37      2      No  
2               376.000 €\nha bajado 20.000 €  376000   60      2      No  
3                                   715.000 €  715000  101      3      No  
4               380.000 €\nha bajado 35.000 €  380000   82      2      No

## Analizamos los valores null

In [17]:
df_Comprar.isnull().sum()

Unnamed: 0      0
id              0
url             0
subtype         0
transaction     0
owner           0
sell-type       0
provider        0
title           0
location        0
features        1
description     0
price          30
price2          0
m2              0
rooms           0
parking         0
dtype: int64

In [18]:
df_Comprar[df_Comprar['price'].isnull()]

Unnamed: 0                id  \
2950         2950  id42001000000006   
6279         6279  id28371003772984   
6280         6280   id8166003793038   
6281         6281   id4975004333174   
6282         6282   id4975004333170   
6283         6283   id4975004333172   
6284         6284  id36906000004377   
6285         6285  id36271004087414   
6286         6286  id39904000000492   
6287         6287  id39904000000493   
6288         6288  id39904000000494   
6299         6299   id6069004160283   
7410         7410   id5115003753948   
7439         7439   id3371004288726   
8881         8881  id12376004119864   
8882         8882  id20418004351603   
8965         8965   id2868000000173   
9632         9632   id8166003732271   
9633         9633   id8166003785678   
11267       11267  id20418004351607   
12509       12509  id28405000000255   
12510       12510  id28405000000251   
12641       12641  id31686000000325   
14747       14747  id23843000002165   
14748       14748  id36562004212584   
17110       17110  id41687000118197   
17111       17111  id15125003639132   
18747       18747   id4975004330907   
18748       18748   id9861003834465   
18749       18749  id19086004315946   

                                                                                                                                                                                                                              url  \
2950                             https://www.habitaclia.com/comprar-atico-con_3_habitaciones_con_ascensor_calefaccion_aire_acondic_dreta_de_l_eixample-barcelona-i42001000000006.htm?pag=27&f=&geo=d&coddists=200&from=list&lo=55   
6279                              https://www.habitaclia.com/comprar-piso-reformado_de_dos_habitaciones_y_un_bano_frente_a_la_fachada_sagrada_familia-barcelona-i28371003772984.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6280                                                https://www.habitaclia.com/comprar-atico-nuevo_con_terraza_en_rambla_catalunya_dreta_de_l_eixample-barcelona-i8166003793038.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6281                           https://www.habitaclia.com/comprar-atico-de_obra_nueva_de_3_dormitorios_con_59m2_terraza_en_venta_e_dreta_de_l_eixample-barcelona-i4975004333174.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6282                           https://www.habitaclia.com/comprar-piso-en_excelentes_condiciones_de_2_dormitorios_con_26m2_terraza_dreta_de_l_eixample-barcelona-i4975004333170.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6283                           https://www.habitaclia.com/comprar-atico-de_obra_nueva_de_3_dormitorios_con_56m2_terraza_en_venta_e_dreta_de_l_eixample-barcelona-i4975004333172.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6284                          https://www.habitaclia.com/comprar-piso-espectacular_en_finca_modernista_catalogada_con_jardin_de_8_dreta_de_l_eixample-barcelona-i36906000004377.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6285                                                              https://www.habitaclia.com/comprar-atico-obra_nueva_l_antiga_esquerra_de_l_eixample-barcelona-i36271004087414.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6286                                                                  https://www.habitaclia.com/comprar-piso-ramblacatalunya_107_dreta_de_l_eixample-barcelona-i39904000000492.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6287                                                                 https://www.habitaclia.com/comprar-atico-ramblacatalunya_107_dreta_de_l_eixample-barcelona-i39904000000493.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6288                                                                 https://www.habitaclia.com/comprar-atico-ramblacatalunya_107_dreta_de_l_eixample-barcelona-i39904000000494.htm?pag=249&f=&geo=d&coddists=200&from=list&lo=55   
6299                                             

Al no tener información sobre el precio, anulamos estos anuncios :

In [19]:
df_Comprar.drop(df_Comprar[df_Comprar['price'].isnull()].index, inplace = True)

In [20]:
df_Comprar.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 18688 entries, 0 to 18752
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   18688 non-null  int64 
 1   id           18688 non-null  object
 2   url          18688 non-null  object
 3   subtype      18688 non-null  object
 4   transaction  18688 non-null  object
 5   owner        18688 non-null  object
 6   sell-type    18688 non-null  object
 7   provider     18688 non-null  object
 8   title        18688 non-null  object
 9   location     18688 non-null  object
 10  features     18688 non-null  object
 11  description  18688 non-null  object
 12  price        18688 non-null  object
 13  price2       18688 non-null  int64 
 14  m2           18688 non-null  int64 
 15  rooms        18688 non-null  int64 
 16  parking      18688 non-null  object
dtypes: int64(4), object(13)
memory usage: 2.6+ MB


In [21]:
df_Comprar.isna().sum()

Unnamed: 0     0
id             0
url            0
subtype        0
transaction    0
owner          0
sell-type      0
provider       0
title          0
location       0
features       0
description    0
price          0
price2         0
m2             0
rooms          0
parking        0
dtype: int64

## Analizamos las viviendas duplicadas

Es posible que una vivienda sea anunciada por varias inmobiliarias, o que incluso una misma inmobiliaria anuncie la misma vivienda en anuncios diferentes :

In [22]:
df_Comprar[df_Comprar.duplicated(subset=['title','location', 'features', 'price', 'price2', 'm2',
       'rooms'],keep=False)].shape

(2472, 17)

In [23]:
df_Comprar[df_Comprar.duplicated(subset=['location', 'features', 'price', 'price2', 'm2',
       'rooms'],keep=False)].shape

(8146, 17)

Un análisis del dataset nos lleva a concluir que en la mayoría de los casos, cuando una vivienda comparte el barrio y las carácterísticas principales (m2, precio, nºhabitaciones, nºbaños) de otra vivienda, se trata en realidad de la misma vivienda, de manera que decidimos eliminar los duplicados y quedarnos con la primera ocurrencia :

In [24]:
df_Comprar.drop_duplicates(subset=['location', 'features', 'price', 'price2', 'm2',
       'rooms'], keep='first',inplace=True)

In [25]:
df_Comprar.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12426 entries, 0 to 18752
Data columns (total 17 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   12426 non-null  int64 
 1   id           12426 non-null  object
 2   url          12426 non-null  object
 3   subtype      12426 non-null  object
 4   transaction  12426 non-null  object
 5   owner        12426 non-null  object
 6   sell-type    12426 non-null  object
 7   provider     12426 non-null  object
 8   title        12426 non-null  object
 9   location     12426 non-null  object
 10  features     12426 non-null  object
 11  description  12426 non-null  object
 12  price        12426 non-null  object
 13  price2       12426 non-null  int64 
 14  m2           12426 non-null  int64 
 15  rooms        12426 non-null  int64 
 16  parking      12426 non-null  object
dtypes: int64(4), object(13)
memory usage: 1.7+ MB


## Creación de nuevas variables

A partir de los datos disponibles en el dataset, nos parece interesante extraer y/o añadir la siguiente información :
- distrito : la variable 'location' representa los barrios de la ciudad, nos interesa crear la variable 'distrito' que incorpora los 10 distritos de la ciudad, ya que muy probablemente la variable 'location' con 74 ocurrencias no sea de aplicación para nuestros modelos de machine learning.
- cambio de precio : es evidente que el hecho de que una vivienda haya bajado de precio tiene un impacto en su precio final, nos planteamos extraer esta información de la variable 'precio'.
- nº de baños : esta información no existe como variable numérica, nos planteamos crearla a partir de la variable 'features'.

### Variable distrito

In [26]:
df_Comprar['location'].nunique()

74

In [27]:
distritos=pd.read_excel('distritos.xlsx')
distritos

location  \
0                                  Barcelona - Baix Guinardó   
1                                    Barcelona - Barceloneta   
2                                  Barcelona - Baró de Viver   
3                             Barcelona - Barri de les Corts   
4                                Barcelona - Besòs - Maresme   
5                                     Barcelona - Bon Pastor   
6                                       Barcelona - Can Baró   
7                                    Barcelona - Can Peguera   
8                                      Barcelona - Canyelles   
9                                         Barcelona - Carmel   
10                              Barcelona - Ciutat Meridiana   
11                             Barcelona - Congrés - Indians   
12  Barcelona - Diagonal Mar i el Front Marítim del Poblenou   
13                           Barcelona - Dreta de l´Eixample   
14            Barcelona - El Camp d´en Grassot i Gràcia Nova   
15                    Barcelona - El Camp de l´Arpa del Clot   
16                                       Barcelona - El Clot   
17                                       Barcelona - El Coll   
18             Barcelona - El Parc i la Llacuna del Poblenou   
19               Barcelona - El Poble Sec - Parc de Montjuïc   
20                            Barcelona - Font de la Guatlla   
21                                    Barcelona - Fort Pienc   
22                                         Barcelona - Gòtic   
23                                      Barcelona - Guinardó   
24                                         Barcelona - Horta   
25                                   Barcelona - Hostafrancs   
26               Barcelona - L´Antiga Esquerra de l´Eixample   
27                                    Barcelona - La Bordeta   
28                                      Barcelona - La Clota   
29                          Barcelona - La Font d´en Fargues   
30                                  Barcelona - La Guineueta   
31                            Barcelona - La Marina del Port   
32                    Barcelona - La Marina del Prat Vermell   
33                Barcelona - La Nova Esquerra de l´Eixample   
34                                    Barcelona - La Sagrera   
35                                      Barcelona - La Salut   
36                                  Barcelona - La Teixonera   
37                           Barcelona - La Verneda - La Pau   
38                 Barcelona - La Vila Olímpica del Poblenou   
39                                       Barcelona - Montbau   
40                                         Barcelona - Navas   
41                                     Barcelona - Pedralbes   
42                                      Barcelona - Poblenou   
43                                         Barcelona - Porta   
44                                   Barcelona - Prosperitat   
45                       Barcelona - Provençals del Poblenou   
46                                Barcelona - Putget - Farró   
47                                         Barcelona - Raval   
48                                      Barcelona - Roquetes   
49                               Barcelona - Sagrada Família   
50                        Barcelona - Sant Andreu de Palomar   
51                                   Barcelona - Sant Antoni   
52                      Barcelona - Sant Genís dels Agudells   
53                       Barcelona - Sant Gervasi - Bonanova   
54                        Barcelona - Sant Gervasi - Galvany   
55                      Barcelona - Sant Martí de Provençals   
56                       Barcelona - Sant Ramon - Maternitat   
57                                         Barcelona - Sants   
58                                   Barcelona - Sants-Badal   
59                                        Barcelona - Sarrià   
60            Barcelona - St. Pere - Sta. Caterina - El Born   
61                                    Barcelona - Torre Baró   
62                  

In [28]:
distritos.distrito.value_counts()

Nou Barris             13
Horta Guinardo         11
Sant Marti             10
Sants Montjuic          9
Sant Andreu             7
Eixample                6
Sarria Sant Gervasi     6
Gracia                  5
Ciutat Vella            4
Les Corts               3
Name: distrito, dtype: int64

In [29]:
df_Comprar=pd.merge(df_Comprar,distritos[['distrito','location']],on='location',how='left')

In [30]:
df_Comprar.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12426 entries, 0 to 12425
Data columns (total 18 columns):
 #   Column       Non-Null Count  Dtype 
---  ------       --------------  ----- 
 0   Unnamed: 0   12426 non-null  int64 
 1   id           12426 non-null  object
 2   url          12426 non-null  object
 3   subtype      12426 non-null  object
 4   transaction  12426 non-null  object
 5   owner        12426 non-null  object
 6   sell-type    12426 non-null  object
 7   provider     12426 non-null  object
 8   title        12426 non-null  object
 9   location     12426 non-null  object
 10  features     12426 non-null  object
 11  description  12426 non-null  object
 12  price        12426 non-null  object
 13  price2       12426 non-null  int64 
 14  m2           12426 non-null  int64 
 15  rooms        12426 non-null  int64 
 16  parking      12426 non-null  object
 17  distrito     12426 non-null  object
dtypes: int64(4), object(14)
memory usage: 1.8+ MB


In [31]:
df_Comprar.distrito.value_counts()

Eixample               2590
Ciutat Vella           1870
Sarria Sant Gervasi    1346
Sant Marti             1313
Sants Montjuic         1264
Horta Guinardo         1032
Nou Barris              966
Sant Andreu             768
Gracia                  735
Les Corts               542
Name: distrito, dtype: int64

### Variable cambio_precio

Este concepto se encuentra en la columna 'price' :

In [32]:
df_Comprar['cambio_precio']= np.where(df_Comprar['price'].str.contains(r'ha bajado')==True,'Si','No')


In [33]:
df_Comprar.cambio_precio.value_counts()

No    8852
Si    3574
Name: cambio_precio, dtype: int64

### Variable Baños

El número de baños aparece en la columna 'features' :

In [34]:
df_Comprar['baños_pre']= np.where(df_Comprar['features'].str.contains(r'habitacion')==True,
                         df_Comprar['features'].str.split('-').str[2],
                        (np.where(df_Comprar['features'].str.contains(r'baño')==True,
                         df_Comprar['features'].str.split('-').str[1],'No')))

In [35]:
df_Comprar['baños_pre'].value_counts()

 1 baño         5767
 2 baños        4505
 3 baños        1202
 4 baños         404
 5 baños         158
No               146
 6 baños          75
 7 baños          46
 8 baños          25
 10 baños          3
 9 baños           3
 5.873€/m2         2
 1.442€/m2         1
 1.198€/m2         1
 1.559€/m2         1
 808€/m2           1
 1.986€/m2         1
 1.691€/m2         1
 2.774€/m2         1
 2.437€/m2         1
 1956 baños        1
 3.274€/m2         1
 2.757€/m2         1
 1.389€/m2         1
 1.587€/m2         1
 1.730€/m2         1
 3.000€/m2         1
 1.506€/m2         1
 2.417€/m2         1
 1.215€/m2         1
 7.251€/m2         1
 4.000€/m2         1
 5.000€/m2         1
 4.447€/m2         1
 2.333€/m2         1
 8.305€/m2         1
 120 baños         1
 5.093€/m2         1
 3.875€/m2         1
 4.700€/m2         1
 2.100€/m2         1
 4.070€/m2         1
 2.806€/m2         1
 2.627€/m2         1
 2.621€/m2         1
 2.667€/m2         1
 3.988€/m2         1
 2.624€/m2   

Asignamos el número de baños que corresponde, en el caso de que no haya valor, asignamos el valor 1 por defecto al considerar que una vivienda tiene como mínimo un baño :

In [36]:
df_Comprar['baños']= np.where(df_Comprar['baños_pre'].str.contains(r'baño')==True,
                          df_Comprar['baños_pre'].str.split(' ').str[1],1)

In [37]:
df_Comprar.baños.value_counts()

1       5767
2       4505
3       1202
4        404
1        235
5        158
6         75
7         46
8         25
9          3
10         3
11         1
1956       1
120        1
Name: baños, dtype: int64

Convertimos la variable a numérica :

In [38]:
df_Comprar['baños'] = df_Comprar['baños'].astype('int')

In [39]:
df_Comprar.drop(['baños_pre'], axis=1, inplace=True)

In [40]:
df_Comprar.drop(['Unnamed: 0'], axis=1, inplace=True)

### Dataset final

In [41]:
df_Comprar.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 12426 entries, 0 to 12425
Data columns (total 19 columns):
 #   Column         Non-Null Count  Dtype 
---  ------         --------------  ----- 
 0   id             12426 non-null  object
 1   url            12426 non-null  object
 2   subtype        12426 non-null  object
 3   transaction    12426 non-null  object
 4   owner          12426 non-null  object
 5   sell-type      12426 non-null  object
 6   provider       12426 non-null  object
 7   title          12426 non-null  object
 8   location       12426 non-null  object
 9   features       12426 non-null  object
 10  description    12426 non-null  object
 11  price          12426 non-null  object
 12  price2         12426 non-null  int64 
 13  m2             12426 non-null  int64 
 14  rooms          12426 non-null  int64 
 15  parking        12426 non-null  object
 16  distrito       12426 non-null  object
 17  cambio_precio  12426 non-null  object
 18  baños          12426 non-n

In [42]:
round(df_Comprar.describe(),2).T

count       mean         std  min       25%       50%       75%  \
price2  12426.0  671035.10  9886390.66  1.0  235000.0  375000.0  680000.0   
m2      12426.0     124.32      850.90  1.0      68.0      89.0     129.0   
rooms   12426.0       2.94        1.49 -1.0       2.0       3.0       4.0   
baños   12426.0       1.94       17.59  1.0       1.0       2.0       2.0   

                 max  
price2  1.100000e+09  
m2      9.400000e+04  
rooms   5.500000e+01  
baños   1.956000e+03

In [43]:
round(df_Comprar.describe(include='all')).T

count unique  \
id               12426  12426   
url              12426  12426   
subtype          12426     12   
transaction      12426      1   
owner            12426      2   
sell-type        12426      1   
provider         12426   1722   
title            12426  10892   
location         12426     74   
features         12426  10531   
description      12426  11179   
price            12426   4023   
price2         12426.0    NaN   
m2             12426.0    NaN   
rooms          12426.0    NaN   
parking          12426      2   
distrito         12426     10   
cambio_precio    12426      2   
baños          12426.0    NaN   

                                                                                                                                                       top  \
id                                                                                                                                        id41022004397639   
url            https://www.habitaclia.com/comprar-piso-obra_nueva_calle_carme_23_raval-barcelona-i41022004397639.htm?f=&geo=d&coddists=100&from=list&lo=55   
subtype                                                                                                                                               FLAT   
transaction                                                                                                                                           SALE   
owner                                                                                                                                         PROFESSIONAL   
sell-type                                                                                                                                      SECOND_HAND   
provider                                                                                                                                            8166-1   
title                                                                                                              Piso en Dreta de l´Eixample. Obra nueva   
location                                                                                                                   Barcelona - Dreta de l´Eixample   
features                                                                                                        67m2 - 3 habitaciones - 1 baño - 3.597€/m2   
description                    Esta promoción deja atrás su antiguo uso como edificio industrial para albergar 32 magníficas viviendas. En el nuevo inm...   
price                                                                                                                                            350.000 €   
price2                                                                                                                                                 NaN   
m2                                                                                                                                                     NaN   
rooms                                                                                                                                                  NaN   
parking                                                                                                                                                 No   
distrito                                                                                                                                          Eixample   
cambio_precio                                                                                                                                           No   
baños                                                                                                                                                  NaN   

                freq      mean        std  min       25%       50%       75%  \
id                 1       NaN        NaN  NaN       NaN       NaN       NaN   
url                1       NaN     

## ANALISIS Y LIMPIEZA DE LAS COLUMNAS NUMERICAS

Analizamos los valores máximos y mínimos de las variables numéricas con tal de localizar outliers en el dataset que se correspondan a errores del anuncio :

In [44]:
df_Comprar.describe().apply(lambda s: s.apply('{0:.5f}'.format))

price2           m2        rooms        baños
count       12426.00000  12426.00000  12426.00000  12426.00000
mean       671035.09689    124.32094      2.94415      1.94270
std       9886390.65713    850.90350      1.48926     17.59347
min             1.00000      1.00000     -1.00000      1.00000
25%        235000.00000     68.00000      2.00000      1.00000
50%        375000.00000     89.00000      3.00000      2.00000
75%        680000.00000    129.00000      4.00000      2.00000
max    1100000000.00000  94000.00000     55.00000   1956.00000

### Variable precio

In [45]:
df_Comprar[df_Comprar['price2']>10000000]['subtype'].value_counts()

FLAT         2
FARMHOUSE    1
HOUSE        1
DUPLEX       1
Name: subtype, dtype: int64

In [46]:
df_Comprar[df_Comprar['price2']>10000000]

id  \
2482    id9873002101909   
4417    id5614004098414   
4666   id14671003720579   
6307   id30100000000243   
12226   id1042004270132   

                                                                                                                                                                                                          url  \
2482   https://www.habitaclia.com/comprar-piso-elegante_propiedad_ce_700m2_con_gran_terraza_en_finca_modernista_dreta_de_l_eixample-barcelona-i9873002101909.htm?pag=42&f=&geo=d&coddists=200&from=list&lo=55   
4417                                                https://www.habitaclia.com/comprar-piso-en_paseo_de_gracia_dreta_de_l_eixample-barcelona-i5614004098414.htm?pag=246&f=&geo=d&coddists=200&from=list&lo=55   
4666                                               https://www.habitaclia.com/comprar-masia-castillo_a_pocos_kms_de_vila_de_gracia-barcelona-i14671003720579.htm?pag=14&f=&geo=d&coddists=600&from=list&lo=55   
6307                    https://www.habitaclia.com/comprar-chalet-con_10_habitaciones_amueblado_con_parking_piscina_calef_pedralbes-barcelona-i30100000000243.htm?pag=5&f=&geo=d&coddists=400&from=list&lo=55   
12226                                   https://www.habitaclia.com/comprar-duplex-exclusivo_atico_en_turo_park_sant_gervasi_galvany-barcelona-i1042004270132.htm?pag=91&f=&geo=d&coddists=500&from=list&lo=55   

         subtype transaction         owner    sell-type provider  \
2482        FLAT        SALE  PROFESSIONAL  SECOND_HAND   9873-1   
4417        FLAT        SALE  PROFESSIONAL  SECOND_HAND   5614-1   
4666   FARMHOUSE        SALE  PROFESSIONAL  SECOND_HAND  14671-1   
6307       HOUSE        SALE  PROFESSIONAL  SECOND_HAND  30100-1   
12226     DUPLEX        SALE  PROFESSIONAL  SECOND_HAND   1042-1   

                                                                                               title  \
2482   Piso en Dreta de l´Eixample. Elegante propiedad ce 700m2 con gran terraza en finca modernista   
4417                                 Piso en Dreta de l´Eixample. Piso en barcelona. paseo de gracia   
4666                                      Masía en Vila de Gràcia. Castillo a pocos kms de barcelona   
6307           Chalet en Pedralbes. Chalet con 10 habitaciones amueblado con parking, piscina, calef   
12226                          Dúplex en Sant Gervasi - Galvany. Exclusivo atico-duplex en turo park   

                                 location  \
2482      Barcelona - Dreta de l´Eixample   
4417      Barcelona - Dreta de l´Eixample   
4666           Barcelona - Vila de Gràcia   
6307                Barcelona - Pedralbes   
12226  Barcelona - Sant Gervasi - Galvany   

                                               features  \
2482      700m2 - 4 habitaciones - 6 baños - 19.714€/m2   
4417    1.000m2 - 5 habitaciones - 4 baños - 12.000€/m2   
4666                                2.630m2 - 4.753€/m2   
6307   2.781m2 - 10 habitaciones - 11 baños - 4.495€/m2   
12226  168m2 - 3 habitaciones - 2 baños - 6.547.619€/m2   

                                                                                                                       description  \
2482   Esta majestuosa propiedad de 700m2 se encuentra ubicado en Paseo de Gracia, lugar donde se encuentran las firmas naciona...   
4417   Piso en Barcelona. Paseo de Gracia. Edificio Modernista Catalogado. Conservación original. Más de 800 m2 por planta. Pre...   
4666   Las dos fincas que comprenden la propiedad, están ubicadas en el Vallés Occidental, concretamente muy cerca de la poblac...   
6307   Le invitamos a que conozca uno de los barrios más exquisitos de la Ciudad Condal, donde el lujo y una situación geográfi...   
12226         GESTIÓN PATRIMONIAL OFRECE Magnífico ÁTICO-DÚPLEX en Johann Sebastian Bach junto Turó Park. A DESTACAR Superficie...   

                 price      price2    m2  rooms parking             distrito  \
2482      13.800.000 €    13800000   700      4      No   

In [47]:
df_Comprar[df_Comprar['price2']==1100000000]

id  \
12226  id1042004270132   

                                                                                                                                                                         url  \
12226  https://www.habitaclia.com/comprar-duplex-exclusivo_atico_en_turo_park_sant_gervasi_galvany-barcelona-i1042004270132.htm?pag=91&f=&geo=d&coddists=500&from=list&lo=55   

      subtype transaction         owner    sell-type provider  \
12226  DUPLEX        SALE  PROFESSIONAL  SECOND_HAND   1042-1   

                                                                       title  \
12226  Dúplex en Sant Gervasi - Galvany. Exclusivo atico-duplex en turo park   

                                 location  \
12226  Barcelona - Sant Gervasi - Galvany   

                                               features  \
12226  168m2 - 3 habitaciones - 2 baños - 6.547.619€/m2   

                                                                                                                description  \
12226  GESTIÓN PATRIMONIAL OFRECE Magnífico ÁTICO-DÚPLEX en Johann Sebastian Bach junto Turó Park. A DESTACAR Superficie...   

                 price      price2   m2  rooms parking             distrito  \
12226  1.100.000.000 €  1100000000  168      3      No  Sarria Sant Gervasi   

      cambio_precio  baños  
12226            No      2

In [48]:
df_Comprar.drop(df_Comprar[df_Comprar['price2']==1100000000].index, inplace = True)

In [49]:
df_Comprar[df_Comprar['price2']<50000]['subtype'].value_counts()

FLAT     15
HOUSE     2
LOFT      1
Name: subtype, dtype: int64

In [50]:
df_Comprar[df_Comprar['price2']<40000]

id  \
1835   id19802003567659   
4396   id45015000014214   
5424   id11657004406165   
6170   id38576000000450   
6213   id28396000237978   
6214   id28396000237670   
7190    id5226004430774   
7675   id28396000237796   
7712   id28396000237750   
7726   id28396000237987   
8122    id5226004430791   
9554   id38992000001498   
10804  id45015000020968   
11074  id28396000237854   

                                                                                                                                                                                                url  \
1835                                         https://www.habitaclia.com/comprar-casa-viviendas_en_venta_en_espana_raval-barcelona-i19802003567659.htm?pag=166&f=&geo=d&coddists=100&from=list&lo=55   
4396                     https://www.habitaclia.com/comprar-piso-en_venta_en_pasaje_juvells_manresa_dreta_de_l_eixample-barcelona-i45015000014214.htm?pag=244&f=&geo=d&coddists=200&from=list&lo=55   
5424                          https://www.habitaclia.com/comprar-piso-trasteros_en_venta_passatge_de_flaugier_6_guinardo-barcelona-i11657004406165.htm?pag=16&f=&geo=d&coddists=700&from=list&lo=55   
6170   https://www.habitaclia.com/comprar-loft-local_en_la_teixonera_vall_hebron_junto_parada_metro_linea_5_la_teixonera-barcelona-i38576000000450.htm?pag=93&f=&geo=d&coddists=700&from=list&lo=55   
6213                                                        https://www.habitaclia.com/comprar-piso-la_font_d_en_fargues-barcelona-i28396000237978.htm?pag=96&f=&geo=d&coddists=700&from=list&lo=55   
6214                                                                https://www.habitaclia.com/comprar-piso-la_teixonera-barcelona-i28396000237670.htm?pag=96&f=&geo=d&coddists=700&from=list&lo=55   
7190                                    https://www.habitaclia.com/comprar-piso-con_4_habitaciones_con_ascensor_canyelles-barcelona-i5226004430774.htm?pag=45&f=&geo=d&coddists=800&from=list&lo=55   
7675                                                           https://www.habitaclia.com/comprar-piso-ciutat_meridiana-barcelona-i28396000237796.htm?pag=104&f=&geo=d&coddists=800&from=list&lo=55   
7712                                                               https://www.habitaclia.com/comprar-piso-la_guineueta-barcelona-i28396000237750.htm?pag=107&f=&geo=d&coddists=800&from=list&lo=55   
7726                                                                 https://www.habitaclia.com/comprar-piso-torre_baro-barcelona-i28396000237987.htm?pag=108&f=&geo=d&coddists=800&from=list&lo=55   
8122                              https://www.habitaclia.com/comprar-piso-con_3_habitaciones_con_ascensor_congres_indians-barcelona-i5226004430791.htm?pag=39&f=&geo=d&coddists=900&from=list&lo=55   
9554                                          https://www.habitaclia.com/comprar-piso-con_3_habitaciones_besos_maresme-barcelona-i38992000001498.htm?pag=122&f=&geo=d&coddists=1000&from=list&lo=55   
10804        https://www.habitaclia.com/comprar-chalet-pareado_en_venta_en_calle_santa_maria_balaguer_lerida_la_bordeta-barcelona-i45015000020968.htm?pag=131&f=&geo=d&coddists=300&from=list&lo=55   
11074                                                                     https://www.habitaclia.com/comprar-piso-sants-barcelona-i28396000237854.htm?pag=156&f=&geo=d&coddists=300&from=list&lo=55   

      subtype transaction         owner    sell-type  provider  \
1835    HOUSE        SALE  PROFESSIONAL  SECOND_HAND   19802-1   
4396     FLAT        SALE  PROFESSIONAL  SECOND_HAND   45015-1   
5424     FLAT        SALE  PROFESSIONAL  SECOND_HAND   11657-1   
6170     LOFT        SALE  PROFESSIONAL  SECOND_HAND   38576-1   
6213     FLAT        SALE  PROFESSIONAL  SECOND_HAND  28396-18   
6214     FLAT        SALE  PROFESSIONAL  SECOND_HAND  28396-18   
7190     FLAT        SALE  PROFESSIONAL  SECOND_HAND    5226-1   
7675     FLAT        SALE  PROFESSIONAL  SECOND_HAND  28396-18   
7712     FLAT        SALE  PROFESSI

In [51]:
df_Comprar.drop(df_Comprar[df_Comprar['price2']<40000].index, inplace = True)

### Variable m2

In [52]:
df_Comprar[df_Comprar['m2']>1000]['subtype'].value_counts()

HOUSE        11
FARMHOUSE     1
FLAT          1
Name: subtype, dtype: int64

In [53]:
df_Comprar[df_Comprar['m2']>3000]

id  \
7728  id28396000227044   

                                                                                                                                  url  \
7728  https://www.habitaclia.com/comprar-piso-prosperitat-barcelona-i28396000227044.htm?pag=108&f=&geo=d&coddists=800&from=list&lo=55   

     subtype transaction         owner    sell-type  provider  \
7728    FLAT        SALE  PROFESSIONAL  SECOND_HAND  28396-18   

                    title                 location          features  \
7728  Piso en Prosperitat  Barcelona - Prosperitat  6.800m2 - 15€/m2   

                                                                                                                    description  \
7728  Piso en Barcelona de 52 metros cuadrados, consta de tres dormitorios y un baño OPORTUNIDAD PARA INVERSORES. Inmueble e...   

          price  price2    m2  rooms parking    distrito cambio_precio  baños  
7728  101.990 €  101990  6800      0      No  Nou Barris            No      1

In [54]:
df_Comprar.drop(df_Comprar[df_Comprar['m2']>3000].index, inplace = True)

In [55]:
df_Comprar[df_Comprar['m2']<10]['subtype'].value_counts()

FLAT    3
Name: subtype, dtype: int64

In [56]:
df_Comprar[df_Comprar['m2']<10]

id  \
1327   id1968004433989   
5155  id30028000216940   
7721  id28396000241978   

                                                                                                                                                                                url  \
1327  https://www.habitaclia.com/comprar-piso-de_1hab_1_cuarto_exterior_junto_a_la_rda_sant_antoni_raval-barcelona-i1968004433989.htm?pag=117&f=&geo=d&coddists=100&from=list&lo=55   
5155               https://www.habitaclia.com/comprar-piso-en_venta_en_gracia_travesera_de_dalt_la_salut-barcelona-i30028000216940.htm?pag=71&f=&geo=d&coddists=600&from=list&lo=55   
7721                                                      https://www.habitaclia.com/comprar-piso-porta-barcelona-i28396000241978.htm?pag=107&f=&geo=d&coddists=800&from=list&lo=55   

     subtype transaction         owner    sell-type  provider  \
1327    FLAT        SALE  PROFESSIONAL  SECOND_HAND    1968-1   
5155    FLAT        SALE  PROFESSIONAL  SECOND_HAND   30028-1   
7721    FLAT        SALE  PROFESSIONAL  SECOND_HAND  28396-18   

                                                                                        title  \
1327  Piso Carrer d'en botella. Piso de 1hab + 1 cuarto exterior junto a la rda. sant antoni.   
5155       Piso en La Salut. Piso en venta en gràcia, barcelona (barcelona) travesera de dalt   
7721                                                                            Piso en Porta   

                  location                                      features  \
1327     Barcelona - Raval      3m2 - 1 habitacion - 1 baño - 31.667€/m2   
5155  Barcelona - La Salut  1m2 - 2 habitaciones - 2 baños - 450.000€/m2   
7721     Barcelona - Porta                             1m2 - 780.000€/m2   

                                                                                                                      description  \
1327  Vivienda en el RAVAL situada en la calle Botella / Sant Antoni Abat. Edificio del año 1975 sin ascensor. 3º de altura re...   
5155  Promoción de 1 piso. Promoción de vivienda en el barrio de Gracia, Barcelona.Exclusivo residencial de viviendas en venta...   
7721  Edificio industrial en Barcelona, Fabra i Puig, a rehabilitar para uso residencial, con edificabilidad de 720 m² de tech...   

          price  price2  m2  rooms parking      distrito cambio_precio  baños  
1327   95.000 €   95000   3      1      No  Ciutat Vella            No      1  
5155  450.000 €  450000   1      2      No        Gracia            No      2  
7721  780.000 €  780000   1      0      No    Nou Barris            No      1

In [57]:
df_Comprar.drop(df_Comprar[df_Comprar['m2']<10].index, inplace = True)

In [58]:
df_Comprar.describe().apply(lambda s: s.apply('{0:.5f}'.format))

price2           m2        rooms        baños
count     12407.00000  12407.00000  12407.00000  12407.00000
mean     583277.06649    116.04909      2.94672      1.94398
std      683862.12426    103.61753      1.48760     17.60690
min       42000.00000     13.00000     -1.00000      1.00000
25%      235000.00000     68.00000      2.00000      1.00000
50%      375000.00000     89.00000      3.00000      2.00000
75%      680000.00000    129.00000      4.00000      2.00000
max    13800000.00000   2781.00000     55.00000   1956.00000

### Variable rooms

In [59]:
df_Comprar[df_Comprar['rooms']>10]['subtype'].value_counts()

FLAT     12
HOUSE     9
Name: subtype, dtype: int64

In [60]:
df_Comprar[df_Comprar['rooms']>20]

id  \
6556   id28371003772588   
9498   id39843000000405   
10662  id39843000000406   

                                                                                                                                                                                      url  \
6556   https://www.habitaclia.com/comprar-piso-pedralbes_amplio_y_luminoso_con_vistas_espectaculares_pedralbes-barcelona-i28371003772588.htm?pag=27&f=&geo=d&coddists=400&from=list&lo=55   
9498                      https://www.habitaclia.com/comprar-piso-a_reformar_en_rambla_del_poblenou_poblenou-barcelona-i39843000000405.htm?pag=115&f=&geo=d&coddists=1000&from=list&lo=55   
10662                                https://www.habitaclia.com/comprar-piso-luminoso_en_sants_montjuic_sants-barcelona-i39843000000406.htm?pag=108&f=&geo=d&coddists=300&from=list&lo=55   

      subtype transaction         owner    sell-type provider  \
6556     FLAT        SALE  PROFESSIONAL  SECOND_HAND  28371-1   
9498     FLAT        SALE  PROFESSIONAL  SECOND_HAND  39843-1   
10662    FLAT        SALE  PROFESSIONAL  SECOND_HAND  39843-1   

                                                                                     title  \
6556   Piso en Pedralbes. Pedralbes. piso amplio y luminoso piso con vistas espectaculares   
9498                                 Piso Poblenou. Piso a reformar en rambla del poblenou   
10662                                        Piso Gayarre. Luminoso piso en sants montjuic   

                    location                                       features  \
6556   Barcelona - Pedralbes  325m2 - 55 habitaciones - 4 baños - 5.538€/m2   
9498    Barcelona - Poblenou    82m2 - 22 habitaciones - 1 baño - 4.268€/m2   
10662      Barcelona - Sants    78m2 - 24 habitaciones - 1 baño - 2.865€/m2   

                                                                                                                       description  \
6556   Piso alto y luminoso de 325 m2 con vistas panorámicas, jardín y piscina comunitaria. Próximo al campus universitario, es...   
9498       ¡Presentamos esta vivienda a REFORMAR en la Rambla de PobleNou! Un piso de 82m2 construidos donde son útiles 78m2 do...   
10662    Presentamos esta bonita vivienda en la zona Sant-Montjuic. Un piso de 73?m2 útiles distribuido de la siguiente forma: ...   

                                 price   price2   m2  rooms parking  \
6556   1.800.000 €\nha bajado 95.000 €  1800000  325     55      Si   
9498                         350.000 €   350000   82     22      No   
10662                        223.500 €   223500   78     24      No   

             distrito cambio_precio  baños  
6556        Les Corts            Si      4  
9498       Sant Marti            No      1  
10662  Sants Montjuic            No      1

In [61]:
df_Comprar.drop(df_Comprar[df_Comprar['rooms']>20].index, inplace = True)

In [62]:
df_Comprar[df_Comprar['rooms']==0]['subtype'].value_counts()

FLAT                           61
STUDIO                         18
HOUSE                           8
GROUND_FLOOR                    6
APARTMENT                       5
LOFT                            5
SINGLE_FAMILY_SEMI_DETACHED     2
PENTHOUSE                       2
DUPLEX                          1
FARMHOUSE                       1
Name: subtype, dtype: int64

In [63]:
df_Comprar[df_Comprar['rooms']==-1]['subtype'].value_counts()

FLAT                           99
LOFT                           12
HOUSE                          12
STUDIO                          8
GROUND_FLOOR                    7
APARTMENT                       2
SINGLE_FAMILY_SEMI_DETACHED     1
PENTHOUSE                       1
PAIRED_HOUSE                    1
DUPLEX                          1
Name: subtype, dtype: int64

In [64]:
df_Comprar.drop(df_Comprar[df_Comprar['rooms']==-1].index, inplace = True)

### Variable baños

In [65]:
df_Comprar[df_Comprar['baños']>11]['subtype'].value_counts()

DUPLEX    1
FLAT      1
Name: subtype, dtype: int64

In [66]:
df_Comprar[df_Comprar['baños']>11]

id  \
7561   id23591004307427   
11852   id9295003949761   

                                                                                                                                                                                    url  \
7561                             https://www.habitaclia.com/comprar-duplex-gran_oportunidad_en_la_zona_porta-barcelona-i23591004307427.htm?pag=91&f=&geo=d&coddists=800&from=list&lo=55   
11852  https://www.habitaclia.com/comprar-piso-en_venta_en_bonanova_a_estrenar_tranquilo_sant_gervasi_galvany-barcelona-i9295003949761.htm?pag=58&f=&geo=d&coddists=500&from=list&lo=55   

      subtype transaction         owner    sell-type provider  \
7561   DUPLEX        SALE  PROFESSIONAL  SECOND_HAND  23591-1   
11852    FLAT        SALE  PROFESSIONAL  SECOND_HAND   9295-1   

                                                                                        title  \
7561                         Dúplex Carrer de l'escultor ordóñez. Gran oportunidad en la zona   
11852  Piso Ronda general mitre. Piso en venta en bonanova - barcelona. a estrenar. tranquilo   

                                 location  \
7561                    Barcelona - Porta   
11852  Barcelona - Sant Gervasi - Galvany   

                                             features  \
7561   99m2 - 3 habitaciones - 1956 baños - 3.788€/m2   
11852     84m2 - 1 habitacion - 120 baños - 5.000€/m2   

                                                                                                                       description  \
7561   ÁTICO DUPLEX de 3 habitaciones, dos dobles y una individual. Cocina con galería sin unificar, pero con posibilidad de un...   
11852     Ocasión en Zona Alta con reforma a estrenar ! Piso en Ronda General Mitre. Ideal parejas. Vivienda de unos 84 m2 cons...   

                               price  price2  m2  rooms parking  \
7561                       375.000 €  375000  99      3      No   
11852  420.000 €\nha bajado 30.000 €  420000  84      1      No   

                  distrito cambio_precio  baños  
7561            Nou Barris            No   1956  
11852  Sarria Sant Gervasi            Si    120

In [67]:
df_Comprar.drop(df_Comprar[df_Comprar['baños']>11].index, inplace = True)

In [68]:
df_Comprar.describe().apply(lambda s: s.apply('{0:.5f}'.format))

price2           m2        rooms        baños
count     12258.00000  12258.00000  12258.00000  12258.00000
mean     585088.38098    116.14309      2.98572      1.78536
std      684477.35044    103.29103      1.32973      1.03706
min       42000.00000     13.00000      0.00000      1.00000
25%      237000.00000     68.00000      2.00000      1.00000
50%      375000.00000     89.00000      3.00000      2.00000
75%      682250.00000    130.00000      4.00000      2.00000
max    13800000.00000   2781.00000     17.00000     11.00000

In [69]:
df_Comprar.to_csv('df_Comprar_final.csv')